src/collapse/perf.rs

```diff
@@ -150,33 +207,209 @@ impl Collapse for Folder {
         }
         None
     }
-}
 
-impl From<Options> for Folder {
-    fn from(mut opt: Options) -> Self {
-        opt.include_pid = opt.include_pid || opt.include_tid;
-        Self {
-            in_event: false,
-            skip_stack: false,
-            stack: VecDeque::default(),
-            cache_line: Vec::default(),
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            pname: String::new(),
-            event_filtering: EventFilterState::None,
-            opt,
-        }
+    #[cfg(test)]
+    fn set_nstacks_per_job(&mut self, n: usize) {
+        self.opt.nstacks_per_job = n;
+    }
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
     }
 }
 
 impl Folder {
-    fn on_line(&mut self, line: &str) {
-        if !self.in_event {
-            self.on_event_line(line)
-        } else {
-            self.on_stack_line(line)
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: BufRead,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+            if reader.read_line(&mut line)? == 0 {
+                return Ok(());
+            }
+            if line.starts_with('#') {
+                continue;
+            }
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else if self.in_event {
+                self.on_stack_line(line);
+            } else {
+                self.on_event_line(line);
+            }
         }
     }
 
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
```

Comment:
This shares a lot of code + structure with the `dtrace` implementation. I'd like to see them share as much as possible. May end up being a macro, I'm not sure? Most of my comments from the `dtrace` impl apply here too.

Reply:
I figured out a way to share "all the things" using traits (no macros). I'm leaving the explanation here purposefully vague as I'd like to be able to change some of the key decisions before I push the code for your review, but I have a working version locally. It's a pretty big commit, though; so, I'm going to back out these big changes from my local code, then finish up the other comments that remain (all of which are now small changes, I believe), then at the end I'll push the large changes to you for review, along with a note in the main thread when it's ready.

All that said, it's probably best that I also merge with master first. I'll look up how to do that ... but, **since using git correctly when it comes to merges and forks and branches and stuff is all still pretty new to me, do you have any hints on how I merge the latest master in here without completely breaking things?** Apologies for leaning on you for this kind of thing, but the couple of times I've worked on inferno are really the **only** times I've ever collaborated with someone else on code before (as you may recall, programming became a hobby of mine only fairly recently). So some of the tools (git) and the etiquette (what to write in these messages, how to break up commits) are all really new to me. More generally, on the "etiquette" side, if I'm ever doing something I shouldn't, please let me know. I've already learned a ton about how git and github though working on this one PR alone. :)

Reply:
P.S. Also, your live-coding streams are basically what taught me rust :).

Reply:
Haha, I appreciate that! And you've been doing a heroic effort here too — this is no small PR. As for what to do, the trick is going to be to _first_ make sure you have no uncommitted changes, and then:
```console
$ git remote add upstream https://github.com/jonhoo/inferno.git
$ git fetch upstream
$ git merge upstream/master
```
At this point, git will tell you that there are a number of conflicts — places where the code has changed on `master` and _also_ on your branch. You can see which files are affected by looking for the red lines in the output of
```console
$ git status
```
For each file with conflicts, if you open it in your editor, you'll see something like:
```
<<<<<<< HEAD
    // code from your branch
=======
    // code from master
>>>>>>> master
```
The code before `=======` is the code that's present on your branch. The code after is the code that's present on `master`. Git is basically telling you to manually reconcile the two. To view the changes that's been made on `master` to a particular file in isolation, run:
```console
$ git diff ...master $FILE
```
The `...` indicates "show only commits that have happened since we last diverged from `master`. Oftentimes it can be easier to figure out how to fix a conflict by starting with _your_ version and then just re-making the changes that have happened on master. You can forcefully reset to your branch's version of a file with
```console
$ git show $FILE > $FILE
```
When you've fixed the conflicts for a file, mark that with
```console
$ git add $FILE
```
When `git status` reports no more conflicting files, finish the merge with
```console
$ git commit
```